### ISSUE TITLE
explicit calls to init() and postinit()?

### ISSUE THREAD

**Issue by cassella** [chapel-lang/chapel#8991]
### Summary of Question

Should it be an error to call ``init()`` and ``postinit()`` explicitly?

Currently it's not.

### Steps to Reproduce

**Source Code:**

```chapel
record R {   
  var x: int;

  proc postinit() {
    writeln("R.postinit() x: ", x);
    x = x + 1;
  }
}

var r = new R();

r.postinit();
r.init();
r.postinit();
```

**Compile command:**

chpl post-postinit.chpl

**Execution command:**

```
fortytwo@magrathea:~/src/chapel (master)$ ./post-postinit -nl 1
R.postinit() x: 0
R.postinit() x: 1
R.postinit() x: 0
```

**Associated Future Test(s):**
None yet

### Configuration Information

```
fortytwo@magrathea:~/src/chapel (master)$ chpl --version
chpl version 1.17.0 pre-release (405f366)
Copyright (c) 2004-2018, Cray Inc.  (See LICENSE file for more details)
```
```
fortytwo@magrathea:~/src/chapel (master)$ printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_AUX_FILESYS: none
```
```
fortytwo@magrathea:~/src/chapel (master)$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
```

**Comment by lydia-duncan** [comment]
Yes, it should be an error to make either of these calls outside of a method of the same name (`postinit` should only be called within a `postinit` in the same inheritance hierarchy, ditto for `init`)

**Comment by bradcray** [comment]
I remember deciding that users shouldn't be able to call `init()` or `deinit()`, but I can never remember the rationale for it...  I suspect that the answer for `init()` is that it invokes initializers rather than assignments and fiddles with the dynamic type (for classes) which isn't something that would make sense for a fully-constructed object.  I come up with a rationale for `deinit()`, though... can someone remind me?

I don't see an obvious reason to prevent calls to postinit() offhand...  Is there an argument I'm forgetting?

**Comment by benharsh** [comment]
I don't know if this argument has been made:

I think most users will write a 'postinit' assuming it will only be run once. If it can be called multiple times, then they have to be defensive and potentially store some state to determine if the postinit has already been called.

**Comment by lydia-duncan** [comment]
I think if the function is intended to be used at multiple times in the object's lifetime, then it should be called something other than `postinit`, and `postinit` should call it.

**Comment by bradcray** [comment]
Though he hasn't commented on this issue specifically, I believe that in the past @mppf has advocated for not supporting user calls to `init()`.

**Comment by mppf** [comment]
Right now, I'd expect it to be an error to explicitly call `init()`, at least in user code, since users should be using `new` to do that.

However I think it's important that we eventually support something analogous to C++ placement-new. This could be achieved by allowing `init()` be called directly or it could be handled with different `new` syntax (which is the choice C++ made).

Either way I'd want us to consider it a program error (whether it is reported or not) to call `init()` on the same record or class more than once.
